great_expectations/core/datasource_dict.py

```diff
@@ -92,17 +94,27 @@ def data(self) -> dict[str, FluentDatasource | BaseDatasource]:  # type: ignore[
 
         return datasources
 
-    @override
-    def __setitem__(self, name: str, ds: FluentDatasource | BaseDatasource) -> None:
+    def set_datasource(
+        self, name: str, ds: FluentDatasource | BaseDatasource
+    ) -> FluentDatasource | None:
```

Comment:
`non-blocking`
overloads for the fluent vs base datasource could help with the typing of the return type.

Reply:
Actually I generic would be better.

```python
T = TypeVar('T', bound=FluentDatasource)

   @overload
    def set_datasource(
        self, name: str, ds: T
    ) -> T:
        ...

    @overload
    def set_datasource(
        self, name: str, ds: BaseDatasource
    ) -> None:
       ...

    def set_datasource(
        self, name: str, ds: T | BaseDatasource
    ) -> T | None:


```

Reply:
Thanks for the detailed suggestion! Addressed in https://github.com/great-expectations/great_expectations/pull/8773/commits/7b417a7499e3e6a00075a633205c44b902168cee.